Stabilization v1

sxzhang25 opened the pull request:
Apply stabilization (interframe mode) for v1 motion photos.

sxzhang25 committed:
Extract stabilization matrices and apply stabilization.

- Integrate protobuf to get stabilization data from motion track for v1 photos.
- Incorporate stabilization matrices into OpenGL pipeline for a single triangle strip.

sxzhang25 committed:
Switch to a single extractor in order to sync seeking in both video and motion tracks.

sxzhang25 committed:
Change from stabilization mode to interframe mode (for v1 motion photos).

sxzhang25 committed:
Add new motion photo v1 files for testing stabilization.

sxzhang25 committed:
Add new motion photo v1 files for testing stabilization.

sxzhang25 committed:
Updated stabilization process.

- Created new HomographyMatrix class for supporting matrices and matrix operations.
- Added conversion from image coordinates to GL coordinates.
- Switched to inverse transforms in interframe stabilization mode.

sxzhang25 committed:
Clean up stabilization for a single strip.

sxzhang25 committed:
Add unit tests for HomographyMatrix.

sxzhang25 committed:
Fix stabilization (apply y-axis flip to texture instead of homography transforms). Add support for turning stabilization off in the motion photo widget.

sxzhang25 committed:
Add Constants class and rename test file.

sxzhang25 committed:
Update Javadoc and increase robustness of stabilization on/off.

sxzhang25 committed:
Merge branch 'open-gl' into stabilization-v1

Update code from pipeline branch.

# Conflicts:
#	MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReader.java
#	MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoWidget.java
#	MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/OutputSurface.java
#	MotionPhotoReader/app/src/test/java/com/google/android/libraries/motionphotoreader/BufferProcessorTest.java

sxzhang25 committed:
Delete unused test files, test activities and layouts, and add all test files into assets folder. Add new proto file for v2 image metadata.

sxzhang25 committed:
Refactor all test constants into Constants class.

sxzhang25 committed:
Add increased unit testing and v2 support in BufferProcessor and update BufferProcessorTest to reflect changes.

- Incorporate support for buffer queue timeouts. Include a doRender flag which only sends the frame to the OpenGL pipeline to render if set to true. A buffer queue timeout will set this flag to false. An invalid sample from the media extractor will also set this flag to false.
- Create a loop to make sure that the motion track is checked (if applicable) in addition to the video track.
- Add a debug mode to the BufferProcessor so that unit tests can test stabilization. If debug mode is on, the stabilization homographies will just be identity matrices for each frame.

sxzhang25 committed:
Refactor constants into separate Constants class. Rewrite input buffer behavior test to verify behavior of input buffers during playback.

sxzhang25 committed:
Delete unused classes.

sxzhang25 committed:
Refactor constants into separate Constants class.

sxzhang25 committed:
Delete all raw files and corresponding method to fetch them.

sxzhang25 committed:
Fix playback speed issues.

sxzhang25 committed:
Have widget loop through asset files on long click.

sxzhang25 committed:
Add further stabilization considerations.

- Find the do_not_stabilize bit and update given stabilization mode based on availability or value of that bit.
- Incorporate v2 stabilization (same as v1 stabilization).
- Move all track selections into a single loop for better modularity.

sxzhang25 committed:
Optimize imports and update API version requirements.

sxzhang25 committed:
Fix homography matrices passed to texture render when stabilization is off.

Signed-off-by: sxzhang25 <zsharonx@gmail.com>

sxzhang25 committed:
Change fields set in constructor to final.

sxzhang25 committed:
Merge branch 'master' of https://github.com/googleinterns/motion-photos-filereader into stabilization-v1

# Conflicts:
#	MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReader.java
#	MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/OutputSurface.java

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/Constants.java:
```diff
@@ -0,0 +1,115 @@
+
+    /* *********************************************************************************************
+     * BufferProcessor.java
+     * ********************************************************************************************/
```
This comment doesn't make sense. You'll want to explain what each of the constants are below.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/Constants.java:
```diff
@@ -0,0 +1,115 @@
+
+    /* *********************************************************************************************
+     * MotionPhotoInfo.java
+     * ********************************************************************************************/
```
avoid comments referring to specific file names. (above and below). Prefer just adding comments on all constants.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/Constants.java:
```diff
@@ -0,0 +1,115 @@
+     * Directory names.
+     * ********************************************************************************************/
+
+    static final String MOTION_PHOTOS_DIR = "motionphotos/";
```
add comment what this is for. Is this just for test cases and sample apps? if so, this constant shouldn't be here.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/Constants.java:
```diff
@@ -0,0 +1,115 @@
+    static final String FILENAME_V2 = "PXL_20200710_061629144.MP.jpg";
+    static final long KEY_DURATION_V2 = 763422;
+    static final int KEY_ROTATION_V2 = 0;
+    static final int VIDEO_OFFSET_V2 = 1317283;
```
keep test constants separate from production constants.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/HomographyMatrix.java:
```diff
@@ -0,0 +1,142 @@
+ */
+class HomographyMatrix {
+
+    private List<Float> matrix = new ArrayList<>();
```
Make matrix final

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/HomographyMatrix.java:
```diff
@@ -0,0 +1,142 @@
+    private List<Float> matrix = new ArrayList<>();
+
+    public HomographyMatrix() {
+        for (float f : IDENTITY) {
+            matrix.add(f);
+        }
+    }
```
```
matrix.addAll(IDENTITY)
```

sxzhang25 replied:
I thought this would work originally too, but it didn't compile. IDENTITY is a float array, is that why? I also tried converting IDENTITY to a collection but that ended up being longer than just adding each element individually.

rjuang replied:
hmm.. how about `Arrays.asList( ... )` ?

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/HomographyMatrix.java:
```diff
@@ -0,0 +1,142 @@
+    }
+
+    public HomographyMatrix(List<Float> matrix) {
+        if (matrix.size() != 9) {
+            throw new RuntimeException("List has incorrect number of elements: " + matrix.size());
+        } else {
```
```
Preconditions.checkArgument(matrix.size() == 9, "Provided matrix must have exactly 9 elements");
this.matrix = matrix;
```

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/HomographyMatrix.java:
```diff
@@ -0,0 +1,142 @@
+    }
+
+    public HomographyMatrix(float[] matrix) {
+        if (matrix.length != 9) {
+            throw new RuntimeException("List has incorrect number of elements: " + matrix.length);
+        } else {
+            for (float f : matrix) {
+                this.matrix.add(f);
+            }
+        }
```
```
public HomographyMatrix(float[] matrix) 
  Preconditions.checkArgument(matrix.size() == 9, "Provided matrix must have exactly 9 elements");
  this.matrix = matrix;
}
```

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/HomographyMatrix.java:
```diff
@@ -0,0 +1,142 @@
+            }
+        }
+        return new HomographyMatrix(scaled);
+
```
nit: delete blank line.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/HomographyMatrix.java:
```diff
@@ -0,0 +1,142 @@
+/**
+ * Represents a 3x3 homography transformation in row-major matrix form.
+ *
+ * TODO: write unit test
+ */
```
unit test?  also add java docs to all public methods.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoInfo.java:
```diff
@@ -106,12 +106,12 @@ private static int getVideoOffset(XMPMeta meta) throws XMPException {
+                    String paddingProperty = V2_XMP_PROP_PREFIX
                             + arrayItemIdx
-                            + "]/Container:Item/Item:Padding";
+                            + V2_XMP_PROP_PADDING_SUFFIX;
```
move these constants to MotionPhotoInfo since they look specific to this class only and aren't used outside thie class.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoInfo.java:
```diff
@@ -206,4 +206,7 @@ public long getDurationUs() {
         return rotation;
     }
+
+    // TODO: Add unit test
```
better java doc

sxzhang25 committed:
Added javadoc.

sxzhang25 committed:
Refactor constants and update unit tests for motion photo version. Add Javadoc for constants classes.

sxzhang25 committed:
Refactor constants and update unit tests for motion photo version. Add Javadoc for constants classes.

rjuang commented on MotionPhotoReader/app/src/androidTest/java/com/google/android/libraries/motionphotoreader/MotionPhotoReaderTest.java:
```diff
@@ -32,18 +35,35 @@
+        try {
+            testMotionPhotosList = assetManager.list("motionphotos");
+        } catch (IOException e) {
+            e.printStackTrace();
```
I think in this case, you'll want to throw the exception and not suppress it right ? Maybe have setup throws IOException ?
```
public void setUp throws IOException() {
   ...
   testMotionPhotosList=assetManager.list("motionphotos");
}
```

If you do want to suppress the exception, I would prefer:
```
 Log.e(TAG, "Error while setting up test", e);
```
over e.printStackTrace.

sxzhang25 replied:
Yeah, throwing it makes more sense. Changed.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/Constants.java:
```diff
@@ -0,0 +1,80 @@
+    static final byte[] OPEN_ARR = "<x:xmpmeta".getBytes();
+    static final byte[] CLOSE_ARR = "</x:xmpmeta>".getBytes();
+
+    /**
+     * Asset directory containing all motion photo files.
+     */
+    static final String MOTION_PHOTOS_DIR = "motionphotos/";
```
I think this is specific to either the sample app or the test . Can we move this constant to that specific class using it ? (and if both, then consider re-defining the constant in each class used)

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoWidget.java:
```diff
@@ -108,12 +111,7 @@ public void surfaceChanged(SurfaceHolder holder, int format, int width, int heig
-                            surfaceWidth,
-                            surfaceHeight
-                    );
+                    reader = MotionPhotoReader.open(file, holder.getSurface(), surfaceWidth, surfaceHeight, stabilizationOn);
```
nit: line length > 100 chars

sxzhang25 committed:
Simplify code.

- Move directory constants into respective classes that use it.
- Change identity matrix from float[] to Float[].
- Fix Javadoc.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/Constants.java:
```diff
@@ -0,0 +1,75 @@
+    /**
+     * Time to wait for a blocking queue to offer buffer information when polled, in microseconds.
+     */
+    static final long TIMEOUT_US = 1000L;
```
these constants and below should be `public static final`  (add public to here and below) ..

package visibility is set by the class visibility.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReader.java:
```diff
@@ -212,43 +189,91 @@ static MotionPhotoReader open(File file,
+        //   a. If the bit exists, set stabilizationOn to true if it was originally true
+        //   b. If the bit does not exist, override stabilizationOn and set it to false
+        // 2. If the bit does not exist, then override stabilizationOn and set it to false
+        int version = motionPhotoInfo.getVersion();
+        if (version == MOTION_PHOTO_V1) {
+            for (int i = 0; i < extractor.getTrackCount(); i++) {
+                MediaFormat format = extractor.getTrackFormat(i);
+                String mime = format.getString(MediaFormat.KEY_MIME);
+                assert mime != null;
+                if (mime.startsWith(MOTION_PHOTO_IMAGE_META_MIMETYPE)) {
+                    Log.d(TAG, "selected image meta track: " + i);
+                    extractor.selectTrack(i);
+                    ByteBuffer inputBuffer = ByteBuffer.allocateDirect((int) extractor.getSampleSize());
+                    int sampleSize = extractor.readSampleData(inputBuffer, 0);
+                    if (sampleSize >= 0) {
+                        // The do_not_stabilize bit is available
+                        ImageMeta.ImageData imageData = ImageMeta.ImageData.parseFrom(inputBuffer);
+                        if (imageData.hasDoNotStabilize()) {
+                            stabilizationOn = stabilizationOn && !imageData.getDoNotStabilize();
+                        } else {
+                            stabilizationOn = false;
+                        }
+                    } else {
+                        // The do_not_stabilize bit is unavailable
+                        stabilizationOn = false;
+                    }
+                    extractor.unselectTrack(i);
+                    break;
+                }
+            }
+        }
```
it would be easier to read if you can extract this into a separate private helper method. Prefer something like 
```
private boolean shouldStabilize(... ) { ... }
```

If it helps, use Android Studio's refactor to automatically determine what function inputs you need (i.e. select block of code, right click, refactor -> extract method) and then tweak the resulting extracted method.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/TextureRender.java:
```diff
@@ -82,21 +87,14 @@
-             1.0f,  1.0f    // top right
-    };
-
     private float[] uMatrix = new float[16];
+    private float[] uStabMatrix = new float[16];
```
these two can probably be made `final` . Note that even though they are final, the contents can still be mutated.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/TextureRender.java:
```diff
@@ -321,21 +297,87 @@ private int compileShader(int type, String shaderSrc) {
+        // 3. Store a pointer to the matrix (mark the matrix as transposed, since GL stores matrices
+        //    in column-major order
+        float[] uStabMatrixInv = new float[16];
+        Matrix.setIdentityM(uStabMatrixInv, 0);
```
Maybe make a helper method in HomographyMatrix?

```
public void to4x4(float[] dstRowMajorArray) {   // or colMajor ?? forget which
}
```

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/TextureRender.java:
```diff
@@ -321,21 +297,87 @@ private int compileShader(int type, String shaderSrc) {
+        Log.d(TAG, "Homography list size: " + homographyList.size());
+        glClear(/* mask = */ GL_COLOR_BUFFER_BIT);
+        for (int i = 0; i < NUM_OF_STRIPS; i++) {
+            drawStrip(/* stripIndex = */ i, homographyList.get(i).convertFromImageToGL(videoWidth, videoHeight));
```
line length exceeds 100 chars

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/WidgetTestActivity.java:
```diff
@@ -11,31 +12,46 @@
+    /**
+     * Asset directory containing all motion photo files.
+     */
+    static final String MOTION_PHOTOS_DIR = "motionphotos/";
```
private static final

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/WidgetTestActivity.java:
```diff
@@ -11,31 +12,46 @@
+        try {
+            testMotionPhotosList = assetManager.list("motionphotos");
+        } catch (IOException e) {
+            e.printStackTrace();
```
```
Log.e(TAG, "Error occurred while trying to fetch assets", e);
```

sxzhang25 committed:
Refactor for modularity.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReader.java:
```diff
@@ -95,49 +98,37 @@ private MotionPhotoReader(File file,
-                surface,
-                surfaceWidth,
-                surfaceHeight,
+                surface, surfaceWidth, surfaceHeight,
```
break this into 1 line per argument

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReader.java:
```diff
@@ -95,49 +98,37 @@ private MotionPhotoReader(File file,
-                surface,
-                /* surfaceWidth = */ 0,
-                /* surfaceHeight = */ 0,
+                surface, /* surfaceWidth = */ 0, /* surfaceHeight = */ 0,
```
break this into 1 line per argument

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReader.java:
```diff
@@ -212,43 +189,60 @@ static MotionPhotoReader open(File file,
      */
-    @RequiresApi(api = 23)
-    private void startRenderThread(MotionPhotoInfo motionPhotoInfo) throws IOException {
+    private void startRenderThread(MotionPhotoInfo motionPhotoInfo, boolean stabilizationOn) throws IOException {
```
move `throws IOException` to new line

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReader.java:
```diff
@@ -278,17 +271,61 @@ public void onOutputFormatChanged(@NonNull MediaCodec codec,
+                if (mime.startsWith(MOTION_PHOTO_IMAGE_META_MIMETYPE)) {
+                    Log.d(TAG, "selected image meta track: " + i);
+                    extractor.selectTrack(i);
+                    ByteBuffer inputBuffer = ByteBuffer.allocateDirect((int) extractor.getSampleSize());
```
move `(int) extractor.getSampleSize()` to its own line

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReader.java:
```diff
@@ -278,17 +271,61 @@ public void onOutputFormatChanged(@NonNull MediaCodec codec,
+     * the video.
+     * @throws com.google.protobuf.InvalidProtocolBufferException if parsing an invalid proto object
+     */
+    private boolean isStabilizationOn(MotionPhotoInfo motionPhotoInfo, boolean stabilizationOn)
```
can we call this isAlreadyStabilized ?  also, the method doesn't need to know if stabilization is enabled or not. That logic should be taken care of by the caller.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReader.java:
```diff
@@ -95,49 +98,37 @@ private MotionPhotoReader(File file,
                                          int surfaceWidth,
-                                         int surfaceHeight) throws IOException, XMPException {
+                                         int surfaceHeight,
+                                         boolean stabilizationOn
```
I'm a bit confused what `stabilizationOn` means here. Can we rename to either:

1. `enableStabilization`   (or `applyStabilization`) -- if this value means we should apply stabilization
2. `isPrestabilized` -- if this value means the video is already stabilized  (not sure why we would need a isPrestabilized input, so I'm guessing it's item 1)

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReader.java:
```diff
@@ -278,17 +271,61 @@ public void onOutputFormatChanged(@NonNull MediaCodec codec,
+                        // The do_not_stabilize bit is available
+                        ImageMeta.ImageData imageData = ImageMeta.ImageData.parseFrom(inputBuffer);
+                        if (imageData.hasDoNotStabilize()) {
+                            stabilizationOn = stabilizationOn && !imageData.getDoNotStabilize();
```
whether this video is prestabilized or not should not be impacted by the stabilization setting.

sxzhang25 committed:
Fix code style and refactor code to find isStabilized bit into separate method.

rjuang commented on MotionPhotoReader/app/src/test/java/com/google/android/libraries/motionphotoreader/TestConstants.java:
```diff
@@ -0,0 +1,43 @@
+package com.google.android.libraries.motionphotoreader;
+
+public class TestConstants {
```
remove public keyword here and add public keyword to constants below

rjuang commented on MotionPhotoReader/app/src/main/res/values/attrs.xml:
```diff
@@ -2,6 +2,7 @@
 <resources>
     <declare-styleable name="MotionPhotoWidget">
         <attr name="autoloop" format="boolean" />
+        <attr name="stabilizationOn" format="boolean" />
```
rename to applyStabilization  (this will probably require changes in a bunch of other places)

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReader.java:
```diff
@@ -40,24 +46,22 @@
-    private MediaExtractor lowResExtractor;
-    private MediaCodec lowResDecoder;
-    private MediaExtractor highResExtractor;
+    private MediaExtractor extractor;
```
can extractor be final ?

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReader.java:
```diff
@@ -278,17 +279,63 @@ public void onOutputFormatChanged(@NonNull MediaCodec codec,
+            for (int i = 0; i < extractor.getTrackCount(); i++) {
+                MediaFormat format = extractor.getTrackFormat(i);
+                String mime = format.getString(MediaFormat.KEY_MIME);
+                assert mime != null;
```
Preconditions.checkNotNull(mime);

rjuang reviewed (changes requested):
I think this should be my last round of comments. everything else looks good!

sxzhang25 committed:
Renaming and small changes.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoWidget.java:
```diff
@@ -33,12 +33,15 @@
 
     /** Customizable attribute fields. */
     private final boolean autoloop;
+    private final boolean stabilizationOn;
 
```
rename stabilizationOn to enableStabilization here and below

sxzhang25 committed:
Renaming and small changes.
